x/payment/spec/docs.md

```diff
@@ -0,0 +1,80 @@
+## Abstract
+
+The payment module is responsible for paying for arbitrary data that will be added to the Celestia blockchain. While the data being submitted can be arbitrary, the exact placement of that data is important for the transaction to be valid. This is why the payment module utilizes a malleated transaction scheme. Malleated transactions allow for users to create a single transaction, that can later be malleated by the block producer to create a variety of different valid transactions that are still signed over by the user. To accomplish this, users to create a single `MsgWirePayForMessage` transaction, which is composed of metadata and signatures for multiple variations of the transaction that will be included onchain. After the transaction is submitted to the network, the block producer selects the appropriate signature and creates a valid `MsgPayForMessage` transaction depending on the square size for that block. This new malleated `MsgPayForMessage` transaction is what ends up onchain. 
+
+## State
+The only state that is modified is the sender’s account balance, via the bank keeper’s `Burn` method.
+
+## Messages
+- [`MsgWirePayForMessage`](https://github.com/celestiaorg/celestia-app/blob/b4c8ebdf35db200a9b99d295a13de01110802af4/x/payment/types/tx.pb.go#L32-L40)
+
+While this transaction is created and signed by the user, it never actually ends up onchain. Instead, it is used to create a new “malleated” transaction that does get included onchain.
+- [`MsgPayForMessage`](https://github.com/celestiaorg/celestia-app/blob/b4c8ebdf35db200a9b99d295a13de01110802af4/x/payment/types/tx.pb.go#L208-L216)
+
+The malleated transaction that is created from metadata contained in the original`MsgWirePayFormessage`. It also burns some of the sender’s funds.
+
+## PreProcessTxs
+The malleation process occurs during the PreProcessTxs step.
+
```

Comment:
Why not add an example of the malleation process ?
If this process stays with Celestia in the long term, it makes sense to put an exact example that shows the exact steps, so the process is transparent and everyone is on the same page.

Reply:
okie dokie https://github.com/celestiaorg/celestia-app/pull/187/commits/7458c62254d7fb44c21d9e7a7d776f3e8cbbec6f